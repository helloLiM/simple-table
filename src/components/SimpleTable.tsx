import { defineComponent, watchEffect, ref } from "vue";
import { TableProps, tableProps } from "./types";
import type { TabData, Column } from "./interface";
import tableHeader from "./table_header";
import tablePagination from "./table_pagination"
import * as _ from 'lodash'
import './style/table.less'

const ASC = "ASC";
const DESC = "DESC";

export default defineComponent({
  name: "SimpleTable",
  props: tableProps,
  components: {
    tableHeader,
    tablePagination
  },
  setup(props: TableProps) {
    let tableData = ref([]); //单页渲染的表格数据
    let allData = ref([]); //所有的表格数据
    let columnData = ref(_.cloneDeep(props.columns)); //表头数据
    let nowPage = ref<number>(1);

    //排序函数
    const sortColumn = (data: Array) => {
      if (!props.sortParams.field) {
        console.warn(`the sort field does not exist`);
        return data;
      }

      if (props.sortParams.direction) {
        console.log(`${props.sortParams.direction} by ${props.sortParams.field}`);

        data.sort((a, b) => {
          const prev = a[props.sortParams.field],
            next = b[props.sortParams.field];
            if(props.sortParams.direction === ASC) {
              return prev < next ? -1 : prev > next ? 1 : 0
            }else if(props.sortParams.direction === DESC){
              return prev < next ? 1 : prev > next ? -1 : 0
            }
        });
      }
      return data;
    };

    // 排序方式传入
    const getDirect = (field: string, direct: ASC | DESC | '') => {
      props.sortParams.field = field;
      props.sortParams.direction = direct;
    };

    //分页函数  接受分页器传入的数据截取开始与结束index
    const changeData = function(start:number, end:number) {
      tableData.value = allData.value.slice(start, end);
    };
    
    
    // 排序监听
    watchEffect(
      () => {
        allData.value = sortColumn(_.cloneDeep(props.tableData));
      }
    );

    // 分页监听
    watchEffect(
      () => {
        let start = props.pageLimit * (nowPage.value - 1);
        tableData.value = allData.value.slice(start, start + props.pageLimit);
      }
    );

    return () => (
      <div>
        <table class="table">
        { props.showHeader ? (<table-header
          columns={columnData.value}
          sort-params={props.sortParams}
          change-sort={getDirect}
          class="table-header"
          ></table-header>) : ''}
        
          <tbody class="table-body">
                {
                  tableData.value.length ?
                  (
                  tableData.value.map((item: TabData, idx) => (
                    <tr  class={'table-tr'} class={'tr' + idx} mark={'row' + item.id}>
                      {
                        props.columns.map((column: Column) => (
                          <td>
                            {
                              item[column.key]
                            }
                          </td>
                        ))
                      }

                    </tr>
                  ))
                  ) : 
                  (
                    <tr>
                      <td colspan={props.columns.length} class="empty-table">
                        暂无数据
                      </td>
                    </tr>
                  )
                }
          </tbody>
        </table>
        <table-pagination
          allData={allData.value}
          page-limit={props.pageLimit}
          page-jump={props.pageJump}
          change-data={changeData}
        ></table-pagination>
      </div>
    );
  },
});
