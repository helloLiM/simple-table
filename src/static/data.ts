export const COLUMN = [{
    key: 'id',
    label: 'ID',
    hidden: 'false',
    sortable: true,
  },{
    key: 'name',
    label: '名称',
    hidden: 'false',
    sortable: true,
  },{
    key: 'sex',
    label: '性别',
    hidden: 'false',
  },{
    key: 'age',
    label: '年龄',
    hidden: 'true',
    sortable: true,
  }]
export const DATA = [
    {
        id: '0',
        name: 'a',
        sex: 'male',
        age: 12,
    }, {
        id: '1',
        name: 'b',
        sex: 'female',
        age: 64,
    }, {
        id: '2',
        name: 'c',
        sex: 'male',
        age: 5,
    }, {
        id: '3',
        name: 'd',
        sex: 'female',
        age: 21,
    }, {
        id: '4',
        name: 'e',
        sex: 'female',
        age: 75,
    }, {
        id: '5',
        name: 'f',
        sex: 'male',
        age: 25,
    }, {
        id: '6',
        name: 'g',
        sex: 'male',
        age: 42,
    }, {
        id: '8',
        name: 'h',
        sex: 'female',
        age: 13,
    },{
        id: '9',
        name: 'wv',
        sex: 'female',
        age: 56,
    },{
        id: '10',
        name: 'ki',
        sex: 'male',
        age: 76,
    },{
        id: '11',
        name: 'xe',
        sex: 'female',
        age: 24,
    },{
        id: '12',
        name: 'yu',
        sex: 'female',
        age: 18,
    },{
        id: '13',
        name: 'sd',
        sex: 'male',
        age: 42,
    },{
        id: '14',
        name: 'vb',
        sex: 'female',
        age: 23,
    },{
        id: '15',
        name: 'dg',
        sex: 'male',
        age: 89,
    },{
        id: '16',
        name: 'nb',
        sex: 'female',
        age: 67,
    },{
        id: '17',
        name: 'fh',
        sex: 'female',
        age: 4,
    },{
        id: '18',
        name: 'ty',
        sex: 'female',
        age: 96,
    },{
        id: '19',
        name: 'ui',
        sex: 'male',
        age: 9,
    },{
        id: '20',
        name: 'mj',
        sex: 'male',
        age: 74,
    },{
        id: '21',
        name: 'pi',
        sex: 'female',
        age: 34,
    },{
        id: '22',
        name: 'ol',
        sex: 'female',
        age: 34,
    },{
        id: '23',
        name: 'asd',
        sex: 'male',
        age: 84,
    },{
        id: '24',
        name: 'az',
        sex: 'male',
        age: 46,
    },{
        id: '25',
        name: 'mu',
        sex: 'female',
        age: 57,
    },{
        id: '26',
        name: 'mn',
        sex: 'male',
        age: 73,
    },{
        id: '27',
        name: 'hg',
        sex: 'female',
        age: 98,
    },{
        id: '28',
        name: 'bg',
        sex: 'female',
        age: 42,
    },{
        id: '29',
        name: 'vcd',
        sex: 'male',
        age: 3,
    },{
        id: '30',
        name: 'qw',
        sex: 'female',
        age: 33,
    },{
        id: '31',
        name: 'y6',
        sex: 'female',
        age: 34,
    },{
        id: '32',
        name: 'wer8',
        sex: 'male',
        age: 42,
    },{
        id: '33',
        name: 'fyj',
        sex: 'female',
        age: 14,
    }
];