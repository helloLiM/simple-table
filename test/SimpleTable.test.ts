import { shallowMount,mount } from "@vue/test-utils";
import { expect, test } from "vitest";
import SimpleTable from "../src/components/SimpleTable";
import { nextTick } from 'vue';
import * as _ from 'lodash';
import { DATA, COLUMN } from "../src/static/data"

test("mount component", () => {
  const wrapper:any = shallowMount(SimpleTable, {
    props: {
      tableData : DATA,
      columns : COLUMN,
      hideHeader : false,
      sortParams :  {
        field: '',
        direction: '',
      },
      pageLimit: 10
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
  
});

// 隐藏表头
test('test hidden table-header', async () => {
  const wrapper:any = shallowMount(SimpleTable, {
    props: {
      tableData : DATA,
      columns : COLUMN,
      showHeader : true,
      sortParams :  {
        field: '',
        direction: '',
      },
      pageLimit: 10
    },
  });
  expect(wrapper.find('.table-header').exists() === true).toBe(true)
  await wrapper.setProps({
    showHeader : false,
  })
  expect(wrapper.find('.table-header').exists() === false).toBe(true)
})

// 排序,模拟传入排序字段
test('test click sort', async () => {
  const sortData = DATA.slice(0, 2);
  const copyData = _.cloneDeep(sortData);

  const wrapper = shallowMount(SimpleTable, {
    props: {
      tableData : copyData,
      columns : COLUMN,
      showHeader : true,
      sortParams :  {
        field: '',
        direction: '',
      },
      pageLimit: 10
    },
  })
  const body:any = wrapper.find('.table-body')
  await wrapper.setProps({
    sortParams :  {
      field: 'id',
      direction: 'ASC',
    },
  })

  expect(body.html().includes('<tr class="table-tr tr0" mark="row0">')).toBe(true)

  await wrapper.setProps({
    sortParams :  {
      field: 'id',
      direction: 'DESC',
    },
  })

  const body1:any = wrapper.find('.table-body')
  expect(body1.html().includes('<tr class="table-tr tr0" mark="row1">')).toBe(true)
})

// 数据为空
test('empty body', async () => {
  const wrapper:any = shallowMount(SimpleTable, {
    props: {
      tableData : DATA,
      columns : COLUMN,
      hideHeader : false,
      sortParams :  {
        field: '',
        direction: '',
      },
      pageLimit: 10
    },
  });
  expect(wrapper.find('.empty-table').exists() === false).toBe(true)
  await wrapper.setProps({
    tableData: [],
  })
  expect(wrapper.find('.empty-table').exists() === true).toBe(true)
})
