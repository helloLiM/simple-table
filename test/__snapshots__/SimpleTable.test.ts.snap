// Vitest Snapshot v1

exports[`mount component 1`] = `
"<div hideheader=\\"false\\">
  <table class=\\"table\\">
    <table-header-stub columns=\\"[object Object],[object Object],[object Object],[object Object]\\" sortparams=\\"[object Object]\\" changesort=\\"(field, direct) => {
      props.sortParams.field = field;
      props.sortParams.direction = direct;
    }\\" class=\\"table-header\\"></table-header-stub>
    <tbody class=\\"table-body\\">
      <tr class=\\"table-tr tr0\\" mark=\\"row0\\">
        <td>0</td>
        <td>a</td>
        <td>male</td>
        <td>12</td>
      </tr>
      <tr class=\\"table-tr tr1\\" mark=\\"row1\\">
        <td>1</td>
        <td>b</td>
        <td>female</td>
        <td>64</td>
      </tr>
      <tr class=\\"table-tr tr2\\" mark=\\"row2\\">
        <td>2</td>
        <td>c</td>
        <td>male</td>
        <td>5</td>
      </tr>
      <tr class=\\"table-tr tr3\\" mark=\\"row3\\">
        <td>3</td>
        <td>d</td>
        <td>female</td>
        <td>21</td>
      </tr>
      <tr class=\\"table-tr tr4\\" mark=\\"row4\\">
        <td>4</td>
        <td>e</td>
        <td>female</td>
        <td>75</td>
      </tr>
      <tr class=\\"table-tr tr5\\" mark=\\"row5\\">
        <td>5</td>
        <td>f</td>
        <td>male</td>
        <td>25</td>
      </tr>
      <tr class=\\"table-tr tr6\\" mark=\\"row6\\">
        <td>6</td>
        <td>g</td>
        <td>male</td>
        <td>42</td>
      </tr>
      <tr class=\\"table-tr tr7\\" mark=\\"row8\\">
        <td>8</td>
        <td>h</td>
        <td>female</td>
        <td>13</td>
      </tr>
      <tr class=\\"table-tr tr8\\" mark=\\"row9\\">
        <td>9</td>
        <td>wv</td>
        <td>female</td>
        <td>56</td>
      </tr>
      <tr class=\\"table-tr tr9\\" mark=\\"row10\\">
        <td>10</td>
        <td>ki</td>
        <td>male</td>
        <td>76</td>
      </tr>
    </tbody>
  </table>
  <table-pagination-stub alldata=\\"[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]\\" pagelimit=\\"10\\" changedata=\\"function (start, end) {
      tableData.value = allData.value.slice(start, end);
    }\\" pagejump=\\"1\\"></table-pagination-stub>
</div>"
`;
